lib/src/main/java/com/knuddels/jtokkit/GptBytePairEncoding.java

```diff
@@ -100,20 +97,14 @@ private EncodingResult encodeOrdinaryInternal(String text, Integer maxTokens) {
         return new EncodingResult(out, false);
     }
 
-    /**
-     * Adds tokens from 'tokensToAdd' to 'out' until either 'maxTokens' is reached or 'tokensToAdd' is exhausted.
-     *
-     * @return the number of tokens added to 'out'
-     */
-    private int addTokens(List<Integer> out, List<Integer> tokensToAdd, Integer maxTokens) {
-        if (maxTokens != null) {
-            List<Integer> sublist = tokensToAdd.subList(0, Math.min(maxTokens - out.size(), tokensToAdd.size()));
-            out.addAll(sublist);
-            return sublist.size();
+    int encodeOrdinaryInternal(String text, Integer maxTokenCount, List<Integer> out) {
+        int tokenCount = 0;
+        List<Integer> ranks = new ArrayList<>(); // reused to avoid allocations
```

Comment:
ranks array is used to keep track of token merges - we initialize it once and reset it before each use to avoid creating so much garbage